File: backend/app/schemas/staff.py
"""Staff schemas — v2 (W1-BE2: スタッフマスタを §4.2 の 9 項目に整理).

設計仕様書 v0.9 §4.2 / 実装手順書 v0.2 §2 W1-BE2 / API 契約 §2 に準拠。

削除済み (v1 → v2): can_double_team / home_address / home_lat / home_lng /
areas / max_per_day / skill_level / assignment_volume の 6 + 3 項目。

状態 (`status`): 在籍 (active) / 休職 (on_leave) / 退職 (retired) の **3 値 enum**
で厳密化 (DB は文字列のまま、Pydantic で Literal バリデート)。

W10-BE1: mentor_id 廃止、is_trainee 追加。
Phase 2 (新人同行 v1.1 §3): 旧 staff_companion_assignments 機構は撤去済み。
同行は trainee_accompaniments (スケジュール画面の同行モード) が正典。
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# 列挙値 (§4.2 / Wave 0-C `app.schemas.v2.staff` と同一定義)
# ---------------------------------------------------------------------------

# 業務ロール (二軸分離・2026-08-09): 実運用は staff(スタッフ) / manager(マネージャー) の 2 値。
# manager は自動割当の対象外 (埋まらないコースの救済にのみ入る)。
# 'admin' は旧データ読み取り互換のためだけに残す (新規入力の選択肢には出さない)。
# ログインのアカウント権限 (users.role: admin/staff) とは **無関係**。
RoleV2 = Literal["admin", "manager", "staff"]

# 状態: 在籍 (active) / 休職 (on_leave) / 退職 (retired) の 3 値
#   保存値は英語、UI 表示は ja-JP に変換する
StaffStatusV2 = Literal["active", "on_leave", "retired"]

# 性別: 患者の性別制限とマッチング用
SexV2 = Literal["male", "female", "unknown"]


class StaffBase(BaseModel):
    """スタッフマスタ v2 基底スキーマ (§4.2 残す 9 項目)."""

    model_config = ConfigDict(extra="forbid")

    # 基本情報
    code: str | None = Field(default=None, max_length=64, description="スタッフコード")
    name: str = Field(min_length=1, max_length=120)
    kana: str | None = Field(default=None, max_length=120)
    sex: SexV2 | None = Field(default=None, description="性別 (患者の性別制限と突合)")
    role: RoleV2 = Field(
        default="staff",
        description="業務ロール。manager は自動割当の対象外 (救済のみ)。アカウント権限とは無関係",
    )
    status: StaffStatusV2 = Field(
        default="active",
        description="在籍 (active) 以外 (on_leave / retired) は割当除外",
    )
    primary_office_id: UUID | None = Field(
        default=None,
        description="主拠点 (§4.3). プルダウン: 稲毛 / 都賀。距離計算の起点",
    )
    note: str | None = Field(default=None, description="自由記述")

    # 詳細セクション (§4.2)
    is_trainee: bool = Field(
        default=False,
        description="新人フラグ。True の場合は同行スタッフ (companion) と一緒に訪問",
    )


class StaffCreate(StaffBase):
    """POST /api/v1/staff リクエスト (W1-BE2)."""


class StaffUpdate(BaseModel):
    """PATCH /api/v1/staff/{id} リクエスト (W1-BE2). 全フィールド optional."""

    model_config = ConfigDict(extra="forbid")

    code: str | None = Field(default=None, max_length=64)
    name: str | None = Field(default=None, min_length=1, max_length=120)
    kana: str | None = Field(default=None, max_length=120)
    sex: SexV2 | None = None
    role: RoleV2 | None = None
    status: StaffStatusV2 | None = None
    primary_office_id: UUID | None = None
    note: str | None = None
    is_trainee: bool | None = None


class StaffRead(StaffBase):
    """GET /api/v1/staff{,/{id}} レスポンス (W1-BE2)."""

    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: UUID
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None = None

File: frontend/lib/schemas/staff.ts
/**
 * Staff zod schemas — mirrors `backend/app/schemas/staff.py` (StaffBase /
 * StaffCreate / StaffUpdate / StaffRead).
 *
 * v2 (W1-BE2 / §4.2): backend は ``extra="forbid"`` で 9 項目のみ受け付ける。
 * 削除済み: ``can_double_team`` / ``home_address`` / ``home_lat`` / ``home_lng`` /
 * ``areas`` / ``max_per_day`` / ``skill_level`` / ``assignment_volume``。
 *
 * 値域:
 *   sex    ∈ male | female | unknown
 *   role   ∈ admin | manager | staff
 *   status ∈ active | on_leave | retired
 *
 * Read-only Wave 1 scope also includes the surrounding domain types referenced
 * by the detail page: weekly shifts, weekly overrides, events, and mentor
 * assignments. Editable mutations for those are deferred to Wave 2 — until then
 * the detail page renders mock data with TODO markers (see below).
 */
import { z } from 'zod';

// ---------------------------------------------------------------------------
// Core staff CRUD payloads (1:1 with backend pydantic models)
// ---------------------------------------------------------------------------

export const STAFF_SEX_VALUES = ['male', 'female', 'unknown'] as const;
export const STAFF_STATUS_VALUES = ['active', 'on_leave', 'retired'] as const;
export const STAFF_ROLE_VALUES = ['admin', 'manager', 'staff'] as const;

export const staffBaseSchema = z.object({
  code: z.string().max(64).nullable().optional(),
  name: z.string().min(1, '氏名は必須です').max(120),
  kana: z.string().max(120).nullable().optional(),
  sex: z.enum(STAFF_SEX_VALUES).nullable().optional(),
  status: z.enum(STAFF_STATUS_VALUES).default('active'),
  role: z.enum(STAFF_ROLE_VALUES).default('staff'),
  primary_office_id: z.string().uuid().nullable().optional(),
  /** Wave 10: 新人スタッフフラグ。true のとき同行スタッフ割付が必要 (§4.2.x) */
  is_trainee: z.boolean().default(false),
  note: z.string().nullable().optional(),
});

export const staffCreateSchema = staffBaseSchema;

export const staffUpdateSchema = z.object({
  code: z.string().max(64).nullable().optional(),
  name: z.string().min(1).max(120).optional(),
  kana: z.string().max(120).nullable().optional(),
  sex: z.enum(STAFF_SEX_VALUES).nullable().optional(),
  status: z.enum(STAFF_STATUS_VALUES).optional(),
  role: z.enum(STAFF_ROLE_VALUES).optional(),
  primary_office_id: z.string().uuid().nullable().optional(),
  /** Wave 10: 新人スタッフフラグ (§4.2.x) */
  is_trainee: z.boolean().optional(),
  note: z.string().nullable().optional(),
});

export const staffReadSchema = staffBaseSchema.extend({
  id: z.string().uuid(),
  created_at: z.string(),
  updated_at: z.string(),
  deleted_at: z.string().nullable().optional(),
});

export type StaffCreate = z.infer<typeof staffCreateSchema>;
export type StaffUpdate = z.infer<typeof staffUpdateSchema>;
export type StaffRead = z.infer<typeof staffReadSchema>;

// ---------------------------------------------------------------------------
// v1 → v2 後方互換 (DB hot-fix 漏れ救済)
// ---------------------------------------------------------------------------

/**
 * 旧 v1 値 (F/M/X や 男/女) を v2 値 (male/female/unknown) に正規化する。
 * 想定外の値が来た場合は ``null`` を返す。
 */
export function normalizeStaffSex(
  v: string | null | undefined,
): (typeof STAFF_SEX_VALUES)[number] | null {
  if (v === null || v === undefined || v === '') return null;
  if (v === 'male' || v === 'female' || v === 'unknown') return v;
  if (['M', 'm', '男', '男性', 'Male'].includes(v)) return 'male';
  if (['F', 'f', '女', '女性', 'Female'].includes(v)) return 'female';
  if (['X', 'x', 'その他', 'Other', 'other'].includes(v)) return 'unknown';
  return null;
}

/**
 * 旧 v1 値 (``inactive``) を v2 値 (``on_leave``) に正規化する。
 * v2 値 (``active`` / ``on_leave`` / ``retired``) はそのまま通す。
 */
export function normalizeStaffStatus(
  v: string | null | undefined,
): (typeof STAFF_STATUS_VALUES)[number] {
  if (v === 'active' || v === 'on_leave' || v === 'retired') return v;
  if (v === 'inactive' || v === '休職') return 'on_leave';
  if (v === '退職') return 'retired';
  return 'active';
}

// ---------------------------------------------------------------------------
// Surrounding domain (read-only Wave 1)
//
// The backend already models these tables (`staff_shifts`,
// `staff_weekly_overrides`, `staff_events`, `mentor_assignments`) but no REST
// endpoints exist yet. Wave 2 will introduce them — until then the detail page
// uses mock fixtures shaped like these schemas.
// ---------------------------------------------------------------------------

/** 0 = Mon ... 6 = Sun (matches backend `staff_shifts.weekday`). */
export const WEEKDAY_LABELS = ['月', '火', '水', '木', '金', '土', '日'] as const;

export const staffShiftSchema = z.object({
  staff_id: z.string().uuid(),
  weekday: z.number().int().min(0).max(6),
  is_on: z.boolean(),
  start_time: z.string().nullable().optional(),
  end_time: z.string().nullable().optional(),
});
export type StaffShift = z.infer<typeof staffShiftSchema>;

export const staffWeeklyOverrideSchema = z.object({
  id: z.string().uuid(),
  staff_id: z.string().uuid(),
  iso_year: z.number().int(),
  iso_week: z.number().int(),
  weekday: z.number().int().min(0).max(6),
  override_type: z.enum(['off', 'custom_time']),
  start_time: z.string().nullable().optional(),
  end_time: z.string().nullable().optional(),
  reason: z.string().nullable().optional(),
});
export type StaffWeeklyOverride = z.infer<typeof staffWeeklyOverrideSchema>;

export const staffEventSchema = z.object({
  id: z.string().uuid(),
  staff_id: z.string().uuid(),
  event_type: z.string(),
  starts_at: z.string(),
  ends_at: z.string(),
  title: z.string().nullable().optional(),
  note: z.string().nullable().optional(),
});
export type StaffEvent = z.infer<typeof staffEventSchema>;

// ---------------------------------------------------------------------------
// Helpers shared by list filters
// ---------------------------------------------------------------------------

export const sexLabel = (sex: StaffRead['sex'] | string | null | undefined): string => {
  // 旧 v1 値 (F/M/X) も後方互換で読めるように normalize 経由で表示
  const normalized = typeof sex === 'string' ? normalizeStaffSex(sex) : (sex ?? null);
  switch (normalized) {
    case 'male':
      return '男性';
    case 'female':
      return '女性';
    case 'unknown':
      return '不明';
    default:
      return '--';
  }
};

export const roleLabel = (role: StaffRead['role']): string => {
  switch (role) {
    case 'admin':
      return '管理者';
    case 'manager':
      return 'マネージャー';
    case 'staff':
      return 'スタッフ';
    default:
      return role ?? '--';
  }
};

export const statusLabel = (status: StaffRead['status'] | string | null | undefined): string => {
  const normalized =
    typeof status === 'string' ? normalizeStaffStatus(status) : (status ?? 'active');
  switch (normalized) {
    case 'active':
      return '在籍';
    case 'on_leave':
      return '休職';
    case 'retired':
      return '退職';
    default:
      return '--';
  }
};
